File: exercises/exercise-1-testing-java/src/main/java/io/miragon/example/application/service/bike/SendRejectionMailService.java
package io.miragon.example.application.service.bike;

import io.miragon.example.application.port.inbound.bike.SendRejectionMailUseCase;
import io.miragon.example.application.port.outbound.bike.BikeSubscriptionRepository;
import io.miragon.example.domain.bike.BikeSubscriptionId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class SendRejectionMailService implements SendRejectionMailUseCase {

    private static final Logger log = LoggerFactory.getLogger(SendRejectionMailService.class);

    private final BikeSubscriptionRepository repository;

    public SendRejectionMailService(BikeSubscriptionRepository repository) {
        this.repository = repository;
    }

    @Override
    public void sendRejectionMail(BikeSubscriptionId subscriptionId) {
        var subscription = repository.find(subscriptionId);
        log.info("Sending rejection mail to {} - bike {} not available",
            subscription.customerEmail().value(), subscription.bikeId().value());
    }

}

File: exercises/exercise-1-testing-kotlin/src/main/kotlin/io/miragon/example/application/service/bike/SendRejectionMailService.kt
package io.miragon.example.application.service.bike

import io.miragon.example.application.port.inbound.bike.SendRejectionMailUseCase
import io.miragon.example.application.port.outbound.bike.BikeSubscriptionRepository
import io.miragon.example.domain.bike.BikeSubscriptionId
import mu.KotlinLogging
import org.springframework.stereotype.Service

@Service
class SendRejectionMailService(
    private val repository: BikeSubscriptionRepository
) : SendRejectionMailUseCase {

    private val log = KotlinLogging.logger {}

    override fun sendRejectionMail(subscriptionId: BikeSubscriptionId) {
        val subscription = repository.find(subscriptionId)
        log.info { "Sending rejection mail to ${subscription.customerEmail.value} - bike ${subscription.bikeId.value} not available" }
    }
}
